x/peggy/abci.go

```diff
@@ -157,58 +157,84 @@ func (h *BlockHandler) cleanupTimedOutBatches(ctx sdk.Context) {
-	if uint64(ctx.BlockHeight()) > params.SignedValsetsWindow {
-		maxHeight = uint64(ctx.BlockHeight()) - params.SignedValsetsWindow
-	} else {
-		// we can't slash anyone if SignedValsetWindow blocks have not passed
+	if uint64(ctx.BlockHeight()) <= params.SignedValsetsWindow {
 		return
 	}
 
-	unslashedValsets := h.k.GetUnslashedValsets(ctx, maxHeight)
+	unslashedValsets := h.k.GetUnslashedValsets(ctx, params.SignedValsetsWindow)
+
+	currentBondedSet := h.k.StakingKeeper.GetBondedValidatorsByPower(ctx)
 
 	// unslashedValsets are sorted by nonce in ASC order
 	for _, vs := range unslashedValsets {
-		confirms := h.k.GetValsetConfirms(ctx, vs.Nonce)
+		confirms := prepValsetConfirms(ctx, h.k, vs.Nonce)
 
 		// SLASH BONDED VALIDATORS who didn't attest valset request
-		currentBondedSet := h.k.StakingKeeper.GetBondedValidatorsByPower(ctx)
-
 		for _, val := range currentBondedSet {
-			consAddr, _ := val.GetConsAddr()
+			consAddr, err := val.GetConsAddr()
+			if err != nil {
+				panic("Failed to get validator consensus addr")
+			}
 			valSigningInfo, exist := h.k.SlashingKeeper.GetValidatorSigningInfo(ctx, consAddr)
 
-			//  Slash validator ONLY if he joined after valset is created
-			if exist && valSigningInfo.StartHeight < int64(vs.Height) {
+			//  Slash validator ONLY if he joined before valset is created
+			startedBeforeValsetCreated := valSigningInfo.StartHeight < int64(vs.Height)
+
+			if exist && startedBeforeValsetCreated {
 				// Check if validator has confirmed valset or not
-				found := false
-				for _, conf := range confirms {
-					ethAddress, exists := h.k.GetEthAddressByValidator(ctx, val.GetOperator())
-					// This may have an issue if the validator changes their eth address
-					// TODO this presents problems for delegate key rotation see issue #344
-					if exists && common.HexToAddress(conf.EthAddress) == ethAddress {
-						found = true
-						break
-					}
-				}
+				_, found := confirms[val.GetOperator().String()]
+
 				// slash validators for not confirming valsets
 				if !found {
-					cons, _ := val.GetConsAddr()
-					consPower := val.ConsensusPower(h.k.StakingKeeper.PowerReduction(ctx))
-
-					h.k.StakingKeeper.Slash(
-						ctx,
-						cons,
-						ctx.BlockHeight(),
-						consPower,
-						params.SlashFractionValset,
-					)
+					// refresh validator before slashing/jailing
+					val, foundVal := h.k.StakingKeeper.GetValidator(ctx, val.GetOperator())
+					if !foundVal {
+						// this should be impossible, we haven't even progressed a single block since we got the list
+						panic("Validator exited set during endblocker?")
+					}
 
 					if !val.IsJailed() {
-						h.k.StakingKeeper.Jail(ctx, cons)
+						consPower := val.ConsensusPower(h.k.StakingKeeper.PowerReduction(ctx))
+						h.k.StakingKeeper.Slash(
+							ctx,
+							consAddr,
+							ctx.BlockHeight(),
+							consPower,
+							params.SlashFractionValset,
+						)
+						ctx.EventManager().EmitEvent(
```

Comment:
see comments below on events 👍

Reply:
This event is still wrong btw

Reply:
@alexanderbez pls check now

Reply:
Nice! Looks good. WDYT about `valset_slashing`?

Reply:
In that case we will need valset_slashing and batch_slashing. Idk how much an event type should englobe, or if it should be very narrow

Reply:
it should be specific to the domain, in this case, valset slashing.

Reply:
Also, we need to update `events.md`, which probably could use an entire overhaul, to reflect these new events.

I don't have a strong opinion on `slashing` vs `valset_slashing`, I just wanted to suggest it.